src/Passes/Source/Rules/Factory.cpp

```diff
@@ -0,0 +1,334 @@
+                auto idx = llvm::APInt(llvm_size.getBitWidth(), llvm_size.getZExtValue() + offset);
+
+                // Computing offset
+                auto new_index = llvm::ConstantInt::get(builder.getContext(), idx);
+
+                // TODO(tfr): Understand what the significance of the addressspace is in relation to the
+                // QIR. Activate by uncommenting:
+                // ptr_type = llvm::PointerType::get(ptr_type->getElementType(), 2);
+                auto instr = new llvm::IntToPtrInst(new_index, ptr_type);
+                instr->takeName(val);
+
+                // Replacing the instruction with new instruction
+                replacements.push_back({llvm::dyn_cast<Instruction>(val), instr});
+
+                // Deleting the getelement and cast operations
+                replacements.push_back({llvm::dyn_cast<Instruction>(cap["getElement"]), nullptr});
+                replacements.push_back({llvm::dyn_cast<Instruction>(cap["cast"]), nullptr});
+
+                return true;
+            };
+
+        auto get_element  = call("__quantum__rt__array_get_element_ptr_1d", "arrayName"_cap = _, "index"_cap = _);
+        auto cast_pattern = bitCast("getElement"_cap = get_element);
+        auto load_pattern = load("cast"_cap = cast_pattern);
+
+        addRule({std::move(load_pattern), access_replacer});
+
+        /// Release replacement
+        auto deleter = deleteInstruction();
+        addRule(
+            {call("__quantum__rt__qubit_release_array", "name"_cap = _),
+             [qubit_alloc_manager, deleter](Builder& builder, Value* val, Captures& cap, Replacements& rep) {
+                 qubit_alloc_manager->release(cap["name"]->getName().str());
+                 return deleter(builder, val, cap, rep);
+             }
+
+            });
+    }
+
+    void RuleFactory::useStaticQubitAllocation()
+    {
+        auto qubit_alloc_manager = qubit_alloc_manager_;
+        auto allocation_replacer =
+            [qubit_alloc_manager](Builder& builder, Value* val, Captures&, Replacements& replacements) {
+                // Getting the type pointer
+                auto ptr_type = llvm::dyn_cast<llvm::PointerType>(val->getType());
+                if (ptr_type == nullptr)
+                {
+                    return false;
+                }
+
+                // Computing the index by getting the current index value and offseting by
+                // the offset at which the qubit array is allocated.
+                auto offset = qubit_alloc_manager->allocate();
+
+                // Creating a new index APInt that is shifted by the offset of the allocation
+                // TODO(tfr): Get the bitwidth size from somewhere
+                auto idx = llvm::APInt(64, offset);
+
+                // Computing offset
+                auto new_index = llvm::ConstantInt::get(builder.getContext(), idx);
+
+                // TODO(tfr): Understand what the significance of the addressspace is in relation to the
```

Comment:
Same as above. There also seems to be some code duplication with the above logic. It would be good to avoid the redundancy.